src/image.js

```diff
@@ -0,0 +1,95 @@
+import * as imageJs from 'image-js';
+import { IOBuffer } from 'iobuffer';
+/**
+ * Reads the data at a certain offset in the file.
+ * It will process the data when I figure it out
+ *
+ * @export
+ * @param {IOBuffer} buffer
+ * @param {number} offset
+ * @param {number} length
+ * @param {number} bytesPerPixel
+ * @return {Number[]}
+ */
+
+export function readPixels(buffer, offset, length, bytesPerPixel) {
+  buffer.offset = offset;
+  const iterations = (length / bytesPerPixel).toFixed(0);
+  const pixels = [];
+  let read;
+  switch (bytesPerPixel) {
+    case 1:
+      read = () => buffer.readChar();
+      break;
+    case 2:
+      read = () => buffer.readUint16();
+      break;
+    case 4:
+      read = () => buffer.readUint32();
+      break;
+    default:
+      throw new Error('Unsupported bytes per pixel resolution');
+  }
+  // Read buffer to fill pixels array
+  for (let i = 0; i < iterations; i++) {
+    pixels[i] = read();
+  }
+  return pixels;
+}
+
+/**
+ * Converts raw height data to actual height values
+ *
+ * @export
+ * @param {object} header
+ * @param {object} image
+ * @return {number}
+ */
+export function dataRescale(header, image) {
+  const zScale = /[^[]*\[(?<softScale>[^\]]*)\] \((?<hardScale>[0-9.]*).*/;
+  let hardScale = 1;
+  let softScale = 1;
+
+  // Scanner list contains soft scale so it needs to be defined
+  if (header['Scanner list'] !== undefined) {
+    if (image['Z scale'] !== undefined) {
+      // Scale attributes also contain units so regex is needed
+      const zScaleExec = zScale.exec(image['Z scale']);
+      softScale = parseFloat(
+        /[^0-9]*(?<scale>[0-9.]*).*/.exec(
+          header['Scanner list'][zScaleExec.groups.softScale],
+        )[1],
+      );
+      hardScale = parseFloat(zScaleExec.groups.hardScale);
+    }
+    // ActualData = hardValue * softScale, hardValue = raw value * hardScale
```

Comment:
I don't understand the following code because map creates a new array that is not used so this code seems useless (and apparently not tested).

'.map' is also quite slow so better to make a simple loop (and change the data inplace ???)
Strange also that you multiplay hardScale * softScale. Is this correct ? If yes there should be precalculated once for ever if they are constant.

Reply:
It is indeed correct and no they should not be constant as they depend on each image contained in the file. I will replace map with a loop, thank you for the input.